sympy/integrals/intpoly.py

```diff
@@ -0,0 +1,806 @@
+        elif p.x/q.x < S.Zero:
+            return S.Zero, p.x * (p.y - q.y)/(q.x - p.x) + p.y
+        else:
+            return ()
+
+    gens = (x, y)
+    power_gens = {}
+
+    for i in gens:
+        power_gens[i] = S.Zero
+
+    if len(gens) > 1:
+        # Special case for vertical and horizontal lines
+        if len(gens) == 2:
+            if a[0] == S.Zero:
+                if y_axis_cut(lineseg):
+                    return S.Zero, b/a[1]
+                else:
+                    return a1, b1
+            elif a[1] == S.Zero:
+                if x_axis_cut(lineseg):
+                    return b/a[0], S.Zero
+                else:
+                    return a1, b1
+
+        if isinstance(expr, Expr):  # Find the sum total of power of each
+            if expr.is_Add:         # generator and store in a dictionary.
+                for monomial in expr.args:
+                    if monomial.is_Pow:
+                        if monomial.args[0] in gens:
+                            power_gens[monomial.args[0]] += monomial.args[1]
+                    else:
+                        for univariate in monomial.args:
+                            term_type = len(univariate.args)
+                            if term_type == 0 and univariate in gens:
+                                power_gens[univariate] += 1
+                            elif term_type == 2 and univariate.args[0] in gens:
+                                power_gens[univariate.args[0]] +=\
+                                           univariate.args[1]
+            elif expr.is_Mul:
+                for term in expr.args:
+                    term_type = len(term.args)
+                    if term_type == 0 and term in gens:
+                        power_gens[term] += 1
+                    elif term_type == 2 and term.args[0] in gens:
+                        power_gens[term.args[0]] += term.args[1]
+            elif expr.is_Pow:
+                power_gens[expr.args[0]] = expr.args[1]
+            elif expr.is_Symbol:
+                power_gens[expr] += 1
+        else:  # If `expr` is a constant take first vertex of the line segment.
+            return a1, b1
+
+        #  TODO : This part is quite hacky. Should be made more robust with
+        #  TODO : respect to symbol names and scalable w.r.t higher dimensions.
+        power_gens = sorted(power_gens.items(), key=lambda k: str(k[0]))
+        if power_gens[0][1] >= power_gens[1][1]:
+            if y_axis_cut(lineseg):
+                x0 = (S.Zero, b / a[1])
+            elif x_axis_cut(lineseg):
+                x0 = (b / a[0], S.Zero)
+            else:
+                x0 = (a1, b1)
+        else:
+            if x_axis_cut(lineseg):
+                x0 = (b/a[0], S.Zero)
+            elif y_axis_cut(lineseg):
+                x0 = (S.Zero, b/a[1])
+            else:
+                x0 = (a1, b1)
+    else:
+        x0 = (b/a[0])
+    return x0
+
+
+def decompose(expr, separate=0):
```

Comment:
I think this `decompose` function could be improved, instead of what is done below could be done with already defined sympy functions like `degree_list`, `degree`.

For a given expression `s`, first extract all symbols. Using `sym = s.atoms(Symbol)` and then, using `degrees = [degree_list(i, *sym), i for i in s.args]` and then building the dictionary using `sum([i for i in degree])` for checking whether the element is already in dictionary and if yes then add the new element to the already existing term, otherwise add a new dictionary `(deg, monom)` pair.

I am not saying that the current logic is wrong, but rather this could be done using already existing functionality.

Apart from that I don't like seeing `if separate:` again and again (unless I don't understand the meaning of `separate`). You could at the end just have `list(poly_dict.values())` (when `separate=True`) would that do the same?